pandas/core/indexes/period.py

```diff
@@ -178,115 +169,111 @@ def __new__(cls, data=None, ordinal=None, freq=None, start=None, end=None,
-
-        if isinstance(data, PeriodIndex):
-            if freq is None or freq == data.freq:  # no freq change
-                freq = data.freq
-                data = data._ndarray_values
+    # ------------------------------------------------------------------------
+    # Mixin Candidates
+    # err. maybe not...
+
+    @property
+    def _box_func(self):
+        def func(x):
+            if isinstance(x, Period) or x is tslib.NaT:
+                return x
             else:
-                base1, _ = _gfc(data.freq)
-                base2, _ = _gfc(freq)
-                data = period.period_asfreq_arr(data._ndarray_values,
-                                                base1, base2, 1)
-            return cls._simple_new(data, name=name, freq=freq)
-
-        # not array / index
-        if not isinstance(data, (np.ndarray, PeriodIndex,
-                                 DatetimeIndex, Int64Index)):
-            if is_scalar(data) or isinstance(data, Period):
-                cls._scalar_data_error(data)
-
-            # other iterable of some kind
-            if not isinstance(data, (list, tuple)):
-                data = list(data)
-
-            data = np.asarray(data)
-
-        # datetime other than period
-        if is_datetime64_dtype(data.dtype):
-            data = dt64arr_to_periodarr(data, freq, tz)
-            return cls._from_ordinals(data, name=name, freq=freq)
-
-        # check not floats
-        if infer_dtype(data) == 'floating' and len(data) > 0:
-            raise TypeError("PeriodIndex does not allow "
-                            "floating point in construction")
-
-        # anything else, likely an array of strings or periods
-        data = ensure_object(data)
-        freq = freq or period.extract_freq(data)
-        data = period.extract_ordinals(data, freq)
-        return cls._from_ordinals(data, name=name, freq=freq)
+                return Period._from_ordinal(ordinal=x, freq=self.freq)
+        return func
+
+    # ------------------------------------------------------------------------
+    # Straight Dispatch
+
+    @property
+    def freq(self):
+        """Return the frequency object if it is set, otherwise None"""
+        return self.values.freq
+
+    # ------------------------------------------------------------------------
+    # Dispatch and Wrap
+
+    def asfreq(self, freq=None, how='E'):
+        result = self.values.asfreq(freq=freq, how=how)
+        return self._simple_new(result, name=self.name)
+    # ------------------------------------------------------------------------
+
 
     @cache_readonly
     def _engine(self):
-        return self._engine_type(lambda: self, len(self))
+        return self._engine_type(lambda: self._ndarray_values, len(self))
+
+    @property
+    def asi8(self):
+        return self.values.asi8
 
     @classmethod
     def _simple_new(cls, values, name=None, freq=None, **kwargs):
+        # TODO: clean up signature.
```

Comment:
Yah, I was thinking this when reading `asfreq` a few lines up.  Maybe a `_from_period_array` or more generally `_from_ea`?

Reply:
In eg IntervalIndex, we kept the name `_simple_new` for the constructor that takes the EA. I think this is also in the same spirit as what `_simple_new` did before (getting the raw ordinals).

I would not call it `_from_period_array`, since then the name will be different for each class.